crates/red_knot_test/src/parser.rs

```diff
@@ -442,45 +587,86 @@ impl<'s> Parser<'s> {
         }
 
         let path = match self.explicit_path {
-            Some(path) => path.to_string(),
-            None => {
-                self.unnamed_file_count += 1;
-
-                match lang {
-                    "py" | "pyi" => format!("mdtest_snippet__{}.{lang}", self.unnamed_file_count),
-                    "" => format!("mdtest_snippet__{}.py", self.unnamed_file_count),
-                    _ => {
-                        bail!(
-                            "Cannot generate name for `lang={}`: Unsupported extension",
-                            lang
-                        );
-                    }
+            Some(path) => {
+                if !EmbeddedFilePath::is_allowed_explicit_path(path) {
+                    bail!("The file name `{path}` must not be used explicitly.",);
                 }
+
+                EmbeddedFilePath::Explicit(path)
             }
+            None => match lang {
+                "py" => EmbeddedFilePath::AutogeneratedPy,
+                "pyi" => EmbeddedFilePath::AutogeneratedPyi,
+                "" => {
+                    bail!("Cannot auto-generate file name for code block with empty language specifier");
+                }
+                _ => {
+                    bail!(
+                        "Cannot auto-generate file name for code block with language `{}`",
+                        lang
+                    );
+                }
+            },
         };
 
-        self.files.push(EmbeddedFile {
-            path: path.clone(),
-            section,
-            lang,
+        let code_block = CodeBlock {
             code,
             backtick_offset,
-        });
+        };
 
-        if let Some(current_files) = &mut self.current_section_files {
-            if !current_files.insert(path.clone()) {
-                bail!(
-                    "Test `{}` has duplicate files named `{path}`.",
-                    self.sections[section].title
-                );
-            };
-        } else {
-            self.current_section_files = Some(FxHashSet::from_iter([path]));
+        let has_merged_snippets = self.current_section_has_merged_snippets();
+        let has_explicit_file_paths = self.current_section_has_explicit_file_paths();
+
+        match self.current_section_files.entry(path.clone()) {
+            Entry::Vacant(entry) => {
+                if has_merged_snippets {
+                    bail!("Merged snippets are not allowed in the presence of other files.");
```

Comment:
Can we make those messages more useful and include some line numbers or at least test name or is this done further up in the parser?